Add an InvalidDocuments exception

Durring the testing of databroker on NSLS2 data we found that occasionally we get a KeyError when filling an event because event['data'].keys() != descriptor['data_keys'].keys(). This is invalid.

I think this we should add an InvalidDocuments exception, so that we have a more meaningful explanation of the problem.

```    
def fill_event(self, doc, include=None, exclude=None, inplace=None):
        try:
            filled = doc['filled']
        except KeyError:
            # This document is not telling us which, if any, keys are filled.
            # Infer that none of the external data is filled.
            descriptor = self._descriptor_cache[doc['descriptor']]
            filled = {key: 'external' in val
                      for key, val in descriptor['data_keys'].items()}
        for key, is_filled in filled.items():
            if exclude is not None and key in exclude:
                continue
            if include is not None and key not in include:
                continue
            if not is_filled:
>               datum_id = doc['data'][key]
E               KeyError: 'Synced_saxs_image'
```

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -645,8 +645,8 @@ class Filler(DocumentRouter):
             filled_doc = doc
         else:
             filled_doc = copy.deepcopy(doc)
-
         descriptor = self._descriptor_cache[doc['descriptor']]
+        from_datakeys = False
         self._current_state.descriptor = descriptor
         try:
             filled = doc['filled']
@@ -655,6 +655,7 @@ class Filler(DocumentRouter):
             # Infer that none of the external data is filled.
             filled = {key: 'external' in val
                       for key, val in descriptor['data_keys'].items()}
+            from_datakeys = True
         for key, is_filled in filled.items():
             self._current_state.key = key
             if exclude is not None and key in exclude:
@@ -662,7 +663,21 @@ class Filler(DocumentRouter):
             if include is not None and key not in include:
                 continue
             if not is_filled:
-                datum_id = doc['data'][key]
+                try:
+                    datum_id = doc['data'][key]
+                except KeyError as err:
+                    if from_datakeys:
+                        raise MismatchedDataKeys(
+￼                            "The documents are not valid.  Either because they "
+￼                            "were recorded incorrectly in the first place, "
+￼                            "corrupted since, or exercising a yet-undiscovered "
+￼                            "bug in reader. event['data'].keys()"
+￼                            "must equal descriptor['data_keys'].keys()."
+￼                            f"event['data'].keys(): {doc['data'].keys()}, "
+￼                            "descriptor['data_keys'].keys(): "
+￼                            f"{descriptor['data_keys'].keys()}") from err
+                    else:
+                        raise err
                 # Look up the cached Datum doc.
                 try:
                     datum_doc = self._datum_cache[datum_id]
@@ -784,13 +799,13 @@ class NoFiller(Filler):
         return filled_doc
 
     def fill_event(self, doc, include=None, exclude=None, inplace=None):
+        descriptor = self._descriptor_cache[doc['descriptor']]
+        from_datakeys = False
         try:
             filled = doc['filled']
-            from_datakeys = False
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.
-            descriptor = self._descriptor_cache[doc['descriptor']]
             filled = {key: 'external' in val
                       for key, val in descriptor['data_keys'].items()}
             from_datakeys = True
@@ -812,7 +827,7 @@ class NoFiller(Filler):
                             "must equal descriptor['data_keys'].keys()."
                             f"event['data'].keys(): {doc['data'].keys()}, "
                             "descriptor['data_keys'].keys(): "
-                            f"{self._descriptor_cache()}") from err
+                            f"{descriptor['data_keys'].keys()}") from err
                     else:
                         raise err
                 # Look up the cached Datum doc.

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -864,8 +864,8 @@ class NoFiller(Filler):
             except KeyError as err:
                 raise UnresolvableForeignKeyError(
                     datum_id,
-                        f"Datum with id {datum_id} refers to unknown Resource "
-                        f"uid {resource_uid}") from err
+                    f"Datum with id {datum_id} refers to unknown Resource "
+                    f"uid {resource_uid}") from err
         return doc
 
 

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -1119,7 +1119,17 @@ class UnresolvableForeignKeyError(EventModelValueError):
 
 
 class DuplicateHandler(EventModelRuntimeError):
-    "raised when a handler is already registered for a given spec"
+    """raised when a handler is already registered for a given spec"""
+    ...
+
+
+class InvalidData(EventModelError):
+    """raised when the data is invalid"""
+    ...
+
+
+class MismatchedDataKeys(InvalidData):
+    """raised when event['data'].keys() != descriptor['data_keys'].keys()"""
     ...
 
 

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -811,7 +811,7 @@ class NoFiller(Filler):
         from_datakeys = False
         try:
             needs_filling = {key for key, val in doc['filled'].items()
-￼                            if val==False}
+                             if val==False}
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.
@@ -823,7 +823,6 @@ class NoFiller(Filler):
                 continue
             if include is not None and key not in include:
                 continue
-            if not is_filled:
             try:
                 datum_id = doc['data'][key]
             except KeyError as err:

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -677,7 +677,15 @@ class Filler(DocumentRouter):
                             "descriptor['data_keys'].keys(): "
                             f"{descriptor['data_keys'].keys()}") from err
                     else:
-                        raise err
+                        raise MismatchedDataKeys(
+                            "The documents are not valid.  Either because they "
+                            "were recorded incorrectly in the first place, "
+                            "corrupted since, or exercising a yet-undiscovered "
+                            "bug in a reader. event['filled'].keys() "
+                            "must be a subset of event['data'].keys(). "
+                            f"event['data'].keys(): {doc['data'].keys()}, "
+                            "event['filled'].keys(): "
+                            f"{doc['filled'].keys()}") from err
                 # Look up the cached Datum doc.
                 try:
                     datum_doc = self._datum_cache[datum_id]
@@ -829,7 +837,15 @@ class NoFiller(Filler):
                             "descriptor['data_keys'].keys(): "
                             f"{descriptor['data_keys'].keys()}") from err
                     else:
-                        raise err
+                        raise MismatchedDataKeys(
+                            "The documents are not valid.  Either because they "
+                            "were recorded incorrectly in the first place, "
+                            "corrupted since, or exercising a yet-undiscovered "
+                            "bug in a reader. event['filled'].keys() "
+                            "must be a subset of event['data'].keys(). "
+                            f"event['data'].keys(): {doc['data'].keys()}, "
+                            "event['filled'].keys(): "
+                            f"{doc['filled'].keys()}") from err
                 # Look up the cached Datum doc.
                 try:
                     datum_doc = self._datum_cache[datum_id]

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -649,75 +649,75 @@ class Filler(DocumentRouter):
         from_datakeys = False
         self._current_state.descriptor = descriptor
         try:
-            filled = doc['filled']
+            needs_filling = {key for key, val in doc['filled'].items()
+                             if val==False}
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.
-            filled = {key: 'external' in val
-                      for key, val in descriptor['data_keys'].items()}
+            needs_filling = {key for key, val in descriptor['data_keys'].items()
+                             if 'external' in val}
             from_datakeys = True
-        for key, is_filled in filled.items():
+        for key in needs_filling:
             self._current_state.key = key
             if exclude is not None and key in exclude:
                 continue
             if include is not None and key not in include:
                 continue
-            if not is_filled:
-                try:
-                    datum_id = doc['data'][key]
-                except KeyError as err:
-                    if from_datakeys:
-                        raise MismatchedDataKeys(
-                            "The documents are not valid.  Either because they "
-                            "were recorded incorrectly in the first place, "
-                            "corrupted since, or exercising a yet-undiscovered "
-                            "bug in a reader. event['data'].keys() "
-                            "must equal descriptor['data_keys'].keys(). "
-                            f"event['data'].keys(): {doc['data'].keys()}, "
-                            "descriptor['data_keys'].keys(): "
-                            f"{descriptor['data_keys'].keys()}") from err
-                    else:
-                        raise MismatchedDataKeys(
-                            "The documents are not valid.  Either because they "
-                            "were recorded incorrectly in the first place, "
-                            "corrupted since, or exercising a yet-undiscovered "
-                            "bug in a reader. event['filled'].keys() "
-                            "must be a subset of event['data'].keys(). "
-                            f"event['data'].keys(): {doc['data'].keys()}, "
-                            "event['filled'].keys(): "
-                            f"{doc['filled'].keys()}") from err
-                # Look up the cached Datum doc.
-                try:
-                    datum_doc = self._datum_cache[datum_id]
-                except KeyError as err:
-                    raise UnresolvableForeignKeyError(
-                        datum_id,
-                        f"Event with uid {doc['uid']} refers to unknown Datum "
-                        f"datum_id {datum_id}") from err
-                resource_uid = datum_doc['resource']
-                # Look up the cached Resource.
-                try:
-                    resource = self._resource_cache[resource_uid]
-                except KeyError as err:
-                    raise UnresolvableForeignKeyError(
-                        resource_uid,
-                        f"Datum with id {datum_id} refers to unknown Resource "
-                        f"uid {resource_uid}") from err
-                handler = self._get_handler_maybe_cached(resource)
-                error_to_raise = DataNotAccessible(
-                        f"Filler was unable to load the data referenced by "
-                        f"the Datum document {datum_doc} and the Resource "
-                        f"document {resource}.")
-                payload = _attempt_with_retries(
-                    func=handler,
-                    args=(),
-                    kwargs=datum_doc['datum_kwargs'],
-                    intervals=[0] + self.retry_intervals,
-                    error_to_catch=IOError,
-                    error_to_raise=error_to_raise)
-                # Here we are intentionally modifying doc in place.
-                filled_doc['data'][key] = payload
-                filled_doc['filled'][key] = datum_id
+            try:
+                datum_id = doc['data'][key]
+            except KeyError as err:
+                if from_datakeys:
+                    raise MismatchedDataKeys(
+                        "The documents are not valid.  Either because they "
+                        "were recorded incorrectly in the first place, "
+                        "corrupted since, or exercising a yet-undiscovered "
+                        "bug in a reader. event['data'].keys() "
+                        "must equal descriptor['data_keys'].keys(). "
+                        f"event['data'].keys(): {doc['data'].keys()}, "
+                        "descriptor['data_keys'].keys(): "
+                        f"{descriptor['data_keys'].keys()}") from err
+                else:
+                    raise MismatchedDataKeys(
+                        "The documents are not valid.  Either because they "
+                        "were recorded incorrectly in the first place, "
+                        "corrupted since, or exercising a yet-undiscovered "
+                        "bug in a reader. event['filled'].keys() "
+                        "must be a subset of event['data'].keys(). "
+                        f"event['data'].keys(): {doc['data'].keys()}, "
+                        "event['filled'].keys(): "
+                        f"{doc['filled'].keys()}") from err
+            # Look up the cached Datum doc.
+            try:
+                datum_doc = self._datum_cache[datum_id]
+            except KeyError as err:
+                raise UnresolvableForeignKeyError(
+                    datum_id,
+                    f"Event with uid {doc['uid']} refers to unknown Datum "
+                    f"datum_id {datum_id}") from err
+            resource_uid = datum_doc['resource']
+            # Look up the cached Resource.
+            try:
+                resource = self._resource_cache[resource_uid]
+            except KeyError as err:
+                raise UnresolvableForeignKeyError(
+                    resource_uid,
+                    f"Datum with id {datum_id} refers to unknown Resource "
+                    f"uid {resource_uid}") from err
+            handler = self._get_handler_maybe_cached(resource)
+            error_to_raise = DataNotAccessible(
+                    f"Filler was unable to load the data referenced by "
+                    f"the Datum document {datum_doc} and the Resource "
+                    f"document {resource}.")
+            payload = _attempt_with_retries(
+                func=handler,
+                args=(),
+                kwargs=datum_doc['datum_kwargs'],
+                intervals=[0] + self.retry_intervals,
+                error_to_catch=IOError,
+                error_to_raise=error_to_raise)
+            # Here we are intentionally modifying doc in place.
+            filled_doc['data'][key] = payload
+            filled_doc['filled'][key] = datum_id
         self._current_state.key = None
         self._current_state.descriptor = None
         return filled_doc

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -786,19 +786,35 @@ class NoFiller(Filler):
     def fill_event(self, doc, include=None, exclude=None, inplace=None):
         try:
             filled = doc['filled']
+            from_datakeys = False
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.
             descriptor = self._descriptor_cache[doc['descriptor']]
             filled = {key: 'external' in val
                       for key, val in descriptor['data_keys'].items()}
+            from_datakeys = True
         for key, is_filled in filled.items():
             if exclude is not None and key in exclude:
                 continue
             if include is not None and key not in include:
                 continue
             if not is_filled:
-                datum_id = doc['data'][key]
+                try:
+                    datum_id = doc['data'][key]
+                except KeyError as err:
+                    if from_datakeys:
+                        raise MismatchedDataKeys(
+                            "The documents are not valid.  Either because they "
+                            "were recorded incorrectly in the first place, "
+                            "corrupted since, or exercising a yet-undiscovered "
+                            "bug in reader. event['data'].keys()"
+                            "must equal descriptor['data_keys'].keys()."
+                            f"event['data'].keys(): {doc['data'].keys()}, "
+                            "descriptor['data_keys'].keys(): "
+                            f"{self._descriptor_cache()}") from err
+                    else:
+                        raise err
                 # Look up the cached Datum doc.
                 try:
                     datum_doc = self._datum_cache[datum_id]

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -650,7 +650,7 @@ class Filler(DocumentRouter):
         self._current_state.descriptor = descriptor
         try:
             needs_filling = {key for key, val in doc['filled'].items()
-                             if val==False}
+                             if val is False}
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.
@@ -811,7 +811,7 @@ class NoFiller(Filler):
         from_datakeys = False
         try:
             needs_filling = {key for key, val in doc['filled'].items()
-                             if val==False}
+                             if val is False}
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -810,59 +810,60 @@ class NoFiller(Filler):
         descriptor = self._descriptor_cache[doc['descriptor']]
         from_datakeys = False
         try:
-            filled = doc['filled']
+            needs_filling = {key for key, val in doc['filled'].items()
+￼                            if val==False}
         except KeyError:
             # This document is not telling us which, if any, keys are filled.
             # Infer that none of the external data is filled.
-            filled = {key: 'external' in val
-                      for key, val in descriptor['data_keys'].items()}
+            needs_filling = {key for key, val in descriptor['data_keys'].items()
+                             if 'external' in val}
             from_datakeys = True
-        for key, is_filled in filled.items():
+        for key in needs_filling:
             if exclude is not None and key in exclude:
                 continue
             if include is not None and key not in include:
                 continue
             if not is_filled:
-                try:
-                    datum_id = doc['data'][key]
-                except KeyError as err:
-                    if from_datakeys:
-                        raise MismatchedDataKeys(
-                            "The documents are not valid.  Either because they "
-                            "were recorded incorrectly in the first place, "
-                            "corrupted since, or exercising a yet-undiscovered "
-                            "bug in a reader. event['data'].keys() "
-                            "must equal descriptor['data_keys'].keys(). "
-                            f"event['data'].keys(): {doc['data'].keys()}, "
-                            "descriptor['data_keys'].keys(): "
-                            f"{descriptor['data_keys'].keys()}") from err
-                    else:
-                        raise MismatchedDataKeys(
-                            "The documents are not valid.  Either because they "
-                            "were recorded incorrectly in the first place, "
-                            "corrupted since, or exercising a yet-undiscovered "
-                            "bug in a reader. event['filled'].keys() "
-                            "must be a subset of event['data'].keys(). "
-                            f"event['data'].keys(): {doc['data'].keys()}, "
-                            "event['filled'].keys(): "
-                            f"{doc['filled'].keys()}") from err
-                # Look up the cached Datum doc.
-                try:
-                    datum_doc = self._datum_cache[datum_id]
-                except KeyError as err:
-                    err_with_key = UnresolvableForeignKeyError(
-                        datum_id,
-                        f"Event with uid {doc['uid']} refers to unknown Datum "
-                        f"datum_id {datum_id}")
-                    err_with_key.key = datum_id
-                    raise err_with_key from err
-                resource_uid = datum_doc['resource']
-                # Look up the cached Resource.
-                try:
-                    self._resource_cache[resource_uid]
-                except KeyError as err:
-                    raise UnresolvableForeignKeyError(
-                        datum_id,
+            try:
+                datum_id = doc['data'][key]
+            except KeyError as err:
+                if from_datakeys:
+                    raise MismatchedDataKeys(
+                        "The documents are not valid.  Either because they "
+                        "were recorded incorrectly in the first place, "
+                        "corrupted since, or exercising a yet-undiscovered "
+                        "bug in a reader. event['data'].keys() "
+                        "must equal descriptor['data_keys'].keys(). "
+                        f"event['data'].keys(): {doc['data'].keys()}, "
+                        "descriptor['data_keys'].keys(): "
+                        f"{descriptor['data_keys'].keys()}") from err
+                else:
+                    raise MismatchedDataKeys(
+                        "The documents are not valid.  Either because they "
+                        "were recorded incorrectly in the first place, "
+                        "corrupted since, or exercising a yet-undiscovered "
+                        "bug in a reader. event['filled'].keys() "
+                        "must be a subset of event['data'].keys(). "
+                        f"event['data'].keys(): {doc['data'].keys()}, "
+                        "event['filled'].keys(): "
+                        f"{doc['filled'].keys()}") from err
+            # Look up the cached Datum doc.
+            try:
+                datum_doc = self._datum_cache[datum_id]
+            except KeyError as err:
+                err_with_key = UnresolvableForeignKeyError(
+                    datum_id,
+                    f"Event with uid {doc['uid']} refers to unknown Datum "
+                    f"datum_id {datum_id}")
+                err_with_key.key = datum_id
+                raise err_with_key from err
+            resource_uid = datum_doc['resource']
+            # Look up the cached Resource.
+            try:
+                self._resource_cache[resource_uid]
+            except KeyError as err:
+                raise UnresolvableForeignKeyError(
+                    datum_id,
                         f"Datum with id {datum_id} refers to unknown Resource "
                         f"uid {resource_uid}") from err
         return doc

--- a/event_model/__init__.py
+++ b/event_model/__init__.py
@@ -668,14 +668,14 @@ class Filler(DocumentRouter):
                 except KeyError as err:
                     if from_datakeys:
                         raise MismatchedDataKeys(
-￼                            "The documents are not valid.  Either because they "
-￼                            "were recorded incorrectly in the first place, "
-￼                            "corrupted since, or exercising a yet-undiscovered "
-￼                            "bug in reader. event['data'].keys()"
-￼                            "must equal descriptor['data_keys'].keys()."
-￼                            f"event['data'].keys(): {doc['data'].keys()}, "
-￼                            "descriptor['data_keys'].keys(): "
-￼                            f"{descriptor['data_keys'].keys()}") from err
+                            "The documents are not valid.  Either because they "
+                            "were recorded incorrectly in the first place, "
+                            "corrupted since, or exercising a yet-undiscovered "
+                            "bug in a reader. event['data'].keys() "
+                            "must equal descriptor['data_keys'].keys(). "
+                            f"event['data'].keys(): {doc['data'].keys()}, "
+                            "descriptor['data_keys'].keys(): "
+                            f"{descriptor['data_keys'].keys()}") from err
                     else:
                         raise err
                 # Look up the cached Datum doc.
@@ -823,8 +823,8 @@ class NoFiller(Filler):
                             "The documents are not valid.  Either because they "
                             "were recorded incorrectly in the first place, "
                             "corrupted since, or exercising a yet-undiscovered "
-                            "bug in reader. event['data'].keys()"
-                            "must equal descriptor['data_keys'].keys()."
+                            "bug in a reader. event['data'].keys() "
+                            "must equal descriptor['data_keys'].keys(). "
                             f"event['data'].keys(): {doc['data'].keys()}, "
                             "descriptor['data_keys'].keys(): "
                             f"{descriptor['data_keys'].keys()}") from err
@@ -1160,7 +1160,11 @@ class InvalidData(EventModelError):
 
 
 class MismatchedDataKeys(InvalidData):
-    """raised when event['data'].keys() != descriptor['data_keys'].keys()"""
+    """
+    Raised when any data keys structures are out of sync. This includes,
+    event['data'].keys(), descriptor['data_keys'].keys(),
+    event['timestamp'].keys(), event['filled'].keys()
+    """
     ...
 
 

--- a/event_model/tests/test_em.py
+++ b/event_model/tests/test_em.py
@@ -953,6 +953,7 @@ def test_filler(tmp_path):
             del event['data']['image']
             filler('event', event)
 
+
 def test_rechunk_event_pages():
 
     def event_page_gen(page_size, num_pages):

--- a/event_model/tests/test_em.py
+++ b/event_model/tests/test_em.py
@@ -931,6 +931,20 @@ def test_filler(tmp_path):
         assert not filler.handler_registry
         assert not filler._handler_cache  # implementation detail
 
+    with pytest.raises(event_model.MismatchedDataKeys):
+        with event_model.Filler(reg) as filler:
+            filler('start', run_bundle.start_doc)
+            descriptor = copy.deepcopy(desc_bundle.descriptor_doc)
+            descriptor['data_keys'] = {}
+            filler('descriptor', descriptor)
+            descriptor = copy.deepcopy(desc_bundle_baseline.descriptor_doc)
+            descriptor['data_keys'] = {}
+            filler('descriptor', descriptor)
+            filler('resource', res_bundle.resource_doc)
+            filler('datum', datum_doc)
+            event = copy.deepcopy(raw_event)
+            event['filled'] = {}
+            filler('event', event)
 
 def test_rechunk_event_pages():
 

--- a/event_model/tests/test_em.py
+++ b/event_model/tests/test_em.py
@@ -932,18 +932,25 @@ def test_filler(tmp_path):
         assert not filler._handler_cache  # implementation detail
 
     with pytest.raises(event_model.MismatchedDataKeys):
-        with event_model.Filler(reg) as filler:
+        with event_model.NoFiller(reg) as filler:
             filler('start', run_bundle.start_doc)
-            descriptor = copy.deepcopy(desc_bundle.descriptor_doc)
-            descriptor['data_keys'] = {}
-            filler('descriptor', descriptor)
-            descriptor = copy.deepcopy(desc_bundle_baseline.descriptor_doc)
-            descriptor['data_keys'] = {}
-            filler('descriptor', descriptor)
+            filler('descriptor', desc_bundle.descriptor_doc)
+            filler('descriptor', desc_bundle_baseline.descriptor_doc)
             filler('resource', res_bundle.resource_doc)
             filler('datum', datum_doc)
             event = copy.deepcopy(raw_event)
-            event['filled'] = {}
+            del event['data']['image']
+            filler('event', event)
+
+    with pytest.raises(event_model.MismatchedDataKeys):
+        with event_model.Filler(reg) as filler:
+            filler('start', run_bundle.start_doc)
+            filler('descriptor', desc_bundle.descriptor_doc)
+            filler('descriptor', desc_bundle_baseline.descriptor_doc)
+            filler('resource', res_bundle.resource_doc)
+            filler('datum', datum_doc)
+            event = copy.deepcopy(raw_event)
+            del event['data']['image']
             filler('event', event)
 
 def test_rechunk_event_pages():

